### PR TITLE
SONARJAVA-867 RSPEC-2165 Finalize fields set check

### PR DESCRIPTION
**Pull request by m-g-sonar**
Finalize fields set check :
- Added new rule to check `null` assignment on fields in `finalize` methods.
- There is no point in setting class fields to `null` in a finalizer. If this this is a hint to the garbage collector, it is unnecessary - the object will be garbage collected anyway - and may actually cause extra work for the garbage collector.

### CODE REVIEW

**Comment by m-g-sonar** [comment]
Code updated with respect to comments from code review.

File: java-checks/src/main/java/org/sonar/java/checks/FinalizeFieldsSetCheck.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    private boolean isFieldAssignment(AssignmentExpressionTree tree) {
+      ExpressionTree variable = tree.variable();
+      if (variable.is(Kind.MEMBER_SELECT)) {
+        variable = ((MemberSelectExpressionTree) variable).expression();
+      }
+      if (variable.is(Kind.IDENTIFIER)) {
+        Symbol variableSymbol = getSemanticModel().getReference((IdentifierTree) variable);
+        return variableSymbol.owner().isKind(Symbol.TYP);
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/FinalizeFieldsSetCheck.java]: This is not sufficient if you have an expression refering a field of another instance  or from another class : 
class A  {
  private int field;
  protected void finalize() {
    new A().field = null;
    new B().field = null;
  }
}

File: java-checks/src/main/java/org/sonar/java/checks/FinalizeFieldsSetCheck.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  @Override
+  public void visitNode(Tree tree) {
+    MethodTree methodTree = (MethodTree) tree;
+    if (isFinalizeDeclaration(methodTree)) {
+      methodTree.accept(new AssignmentVisitor());
+    }
+  }
+
+  private boolean isFinalizeDeclaration(MethodTree tree) {
+    return hasSemantic() && isMethodNamedFinalize(tree) && isOverriding(tree) && hasNoParameters(tree);
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/FinalizeFieldsSetCheck.java]: Check for overriding is uncessary.

File: java-checks/src/main/java/org/sonar/java/checks/FinalizeFieldsSetCheck.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    private boolean isNullAssignment(AssignmentExpressionTree tree) {
+      return tree.expression().is(Kind.NULL_LITERAL);
+    }
+
+    private String getFieldName(AssignmentExpressionTree tree) {
+      ExpressionTree variable = tree.variable();
+      if (variable.is(Kind.MEMBER_SELECT)) {
+        variable = ((MemberSelectExpressionTree) variable).expression();
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/FinalizeFieldsSetCheck.java]: The expression won't contain the field name.

File: java-checks/src/main/java/org/sonar/java/checks/FinalizeFieldsSetCheck.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  @Override
+  public void visitNode(Tree tree) {
+    MethodTree methodTree = (MethodTree) tree;
+    if (isFinalizeDeclaration(methodTree)) {
+      methodTree.accept(new AssignmentVisitor());
+    }
+  }
+
+  private boolean isFinalizeDeclaration(MethodTree tree) {
+    return hasSemantic() && isMethodNamedFinalize(tree) && hasNoParameters(tree);
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/FinalizeFieldsSetCheck.java]: hasSemantic has nothing to do in this method directly.

**Comment by benzonico** [comment]
LGTM

**Pull request closed without merging**